### ISSUE TITLE
高斯噪声

### ISSUE THREAD

**Issue by Singularity0216** [ASCIIJK/MiN-NeurIPS2025#4]
作者您好，这是你提供的PiNoise的代码，请问这里为什么没有涉及高斯噪声，并且还额外加了一个论文中没有提到的x1？

` class PiNoise(torch.nn.Linear):
    def __init__(self, in_dim, out_dim, hidden_dim=384):
        super(torch.nn.Linear, self).__init__()

        self.bias = None

        self.MLP = nn.Linear(in_dim, out_dim)
        torch.nn.init.constant_(self.MLP.weight, 0)
        torch.nn.init.constant_(self.MLP.bias, 0)

        device = torch.device("cuda:{}".format(0))
        factory_kwargs = {"device": device, "dtype": torch.float}

        self.hidden_dim = hidden_dim

        self.w_down = torch.empty((in_dim, self.hidden_dim), **factory_kwargs)
        self.register_buffer("weight", self.w_down)

        self.reset_parameters()

        self.act = nn.GELU()

        self.mu = nn.ModuleList()
        self.sigmma = nn.ModuleList()

        self.w_up = torch.empty((self.hidden_dim, out_dim), **factory_kwargs)
        self.register_buffer("weight", self.w_up)
        self.reset_parameters()

        self.weight_noise = None

    def update_noise(self):

        if len(self.mu) == 0:
            self.mu.append(nn.Linear(self.hidden_dim, self.hidden_dim))
            torch.nn.init.constant_(self.mu[0].weight, 0.)
            torch.nn.init.constant_(self.mu[0].bias, 0.)
            self.sigmma.append(nn.Linear(self.hidden_dim, self.hidden_dim))
            torch.nn.init.constant_(self.sigmma[0].weight, 0.)
            torch.nn.init.constant_(self.sigmma[0].bias, 0.)
        else:
            self.mu.append(nn.Linear(self.hidden_dim, self.hidden_dim))
            torch.nn.init.constant_(self.mu[-1].weight, 0.)
            torch.nn.init.constant_(self.mu[-1].bias, 0.)
            self.sigmma.append(nn.Linear(self.hidden_dim, self.hidden_dim))
            torch.nn.init.constant_(self.sigmma[-1].weight, 0.)
            torch.nn.init.constant_(self.sigmma[-1].bias, 0.)
    
    def init_weight_noise(self, prototypes):
        if len(prototypes) <= 1:
            self.weight_noise = torch.zeros(len(self.mu), requires_grad=True)
        else:
            self.weight_noise = torch.zeros(len(self.mu), requires_grad=True)
            weight = torch.ones(len(self.mu))
            for i in range(len(prototypes)):
                mu_t = prototypes[-1]
                mu_i = prototypes[i]
                dot_product = torch.dot(mu_t, mu_i)
                norm_t = torch.norm(mu_t)
                norm_i = torch.norm(mu_i)
                s_i = dot_product / (norm_t * norm_i)
                weight[i] = s_i.detach().clone()
            weight = torch.softmax(weight, dim=-1)
            self.weight_noise = weight
            self.weight_noise.requires_grad = True
            
    def unfreeze_noise(self):

        for param in self.mu[-1].parameters():
            param.requires_grad = True
        for param in self.sigmma[-1].parameters():
            param.requires_grad = True

    def forward(self, hyper_features):
        x1 = self.MLP(hyper_features)

        x_down = hyper_features @ self.w_down

        noise = None

        for i in range(len(self.mu)):
            mu = self.mu[i](x_down)
            sigmma = self.sigmma[i](x_down)
            if noise is None:
                noise = (mu + sigmma) * self.weight_noise[i]
            else:
                noise += (mu + sigmma) * self.weight_noise[i]

        noise = noise @ self.w_up

        return x1 + noise + hyper_features

    def forward_new(self, hyper_features):
        x1 = self.MLP(hyper_features)

        x_down = hyper_features @ self.w_down

        mu = self.mu[-1](x_down)
        sigmma = self.sigmma[-1](x_down)

        noise = (mu + sigmma) * self.weight_noise[-1]

        noise = noise @ self.w_up

        return x1 + noise + hyper_features`
